src/gpdma.rs

```diff
@@ -0,0 +1,550 @@
+        // width and packing mode is used, therefore the maximum size must be dictated by
+        // destination size (width * count). When not in packing mode, this still holds true as
+        // the destination size must not be exceeded (so only read the same number of words from
+        // the source as there is room in the destination)
+        let size = if src_width > dest_width {
+            dest_size
+        } else {
+            // When the source data width is less than or equal to the destination data width, we
+            // just need to ensure that the destination buffer is large enough to hold all of the
+            // source data.
+            assert!(src_size <= dest_size, "Transfer size ({src_size} bytes) will overflow the destination buffer ({dest_size} bytes)!");
+            src_size
+        };
+
+        // We also need to ensure that the destination
+
+        Self::new::<S::Word, D::Word, MemoryToMemory>(
+            channel, config, src_ptr, dest_ptr, size,
+        )
+    }
+
+    /// Create a new memory-to-peripheral transfer with the channel, source buffer and destination
+    /// peripheral provided.
+    pub fn memory_to_peripheral<S, D>(
+        config: DmaConfig<MemoryToPeripheral, S::Word, D::Word>,
+        channel: &'a CH,
+        source: S,
+        mut destination: D,
+    ) -> Self
+    where
+        S: ReadBuffer<Word: Word>,
+        D: WriteBuffer<Word: Word>,
+    {
+        let (src_ptr, src_words) = unsafe { source.read_buffer() };
+        let src_size = core::mem::size_of::<S::Word>() * src_words;
+        let (dest_ptr, _) = unsafe { destination.write_buffer() };
+
+        Self::new::<S::Word, D::Word, MemoryToPeripheral>(
+            channel, config, src_ptr, dest_ptr, src_size,
+        )
+        .apply_hardware_request_config(config)
+    }
+
+    /// Create a new peripheral-to-memory transfer with the channel, source peripheral and
+    /// destination buffer provided.
+    pub fn peripheral_to_memory<S, D>(
+        config: DmaConfig<PeripheralToMemory, S::Word, D::Word>,
+        channel: &'a CH,
+        source: S,
+        mut destination: D,
+    ) -> Self
+    where
+        S: ReadBuffer<Word: Word>,
+        D: WriteBuffer<Word: Word>,
+    {
+        let (src_ptr, _) = unsafe { source.read_buffer() };
+
+        let (dest_ptr, dest_words) = unsafe { destination.write_buffer() };
+        let dest_size = core::mem::size_of::<D::Word>() * dest_words;
+
+        Self::new::<S::Word, D::Word, PeripheralToMemory>(
+            channel, config, src_ptr, dest_ptr, dest_size,
+        )
+        .apply_hardware_request_config(config)
+        .apply_peripheral_source_config(config)
+    }
+
+    /// Create a new peripheral-to-peripheral transfer with source and destination peripherals
+    /// provided.
+    pub fn peripheral_to_peripheral<S, D, T>(
+        config: DmaConfig<PeripheralToPeripheral<T>, S::Word, D::Word>,
+        channel: &'a CH,
```

Comment:
Is this really ok to have a shared reference? Wouldn't that allow multiple transfers to write to the same channels registers?